Using pem file for automated tsh scripts runs into nil dereference

### Description
I try to copy a file via tsh scp from a CI/CD to a node. The pem file has been certified from the proxy.

**What happened**:
I get a nil dereference error.

**What you expected to happen**:
It should copy a the file to the node behind the proxy.

**How to reproduce it (as minimally and precisely as possible)**:
* Setup proxy with proxy recording
* Generate pem file for second user
* Sign pem file at proxy
* Use pem file to do anything

### Environment
- Teleport version: 4.2.9 (tested with the 4.3.0-alpha2 as well)
- Tsh version: 4.2.9 (tested with the 4.3.0-alpha2 as well)
- OS: macos 10.15 and debian 10
- Where are you running Teleport?: Proxy running on a centos vm at hetzner. tsh in a Bitbucket Cloud Pipeline

**Relevant Debug Logs If Applicable**

- `tsh --debug -i my.pem --proxy teleport.example.com scp localfile root@node:remotefile`
```
INFO             Reading identity file: my.pem common/identity.go:44
DEBU [TSH]       Extracted username "user@domain.tld" from the identity file my.pem. tsh/tsh.go:967
INFO [CLIENT]    Connecting proxy=teleport.example.com:3023 login='root' method=0 client/api.go:1603
INFO [CLIENT]    Successful auth with proxy teleport.example.com:3023 client/api.go:1585
DEBU [CLIENT]    Found clusters: [{"name":"mycluster","lastconnected":"2020-07-06T01:11:47.919662751+02:00","status":"online"}] client/client.go:106
INFO [CLIENT]    Client= connecting to node=node on cluster mycluster client/client.go:573
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x485712e]

goroutine 1 [running]:
github.com/gravitational/teleport/lib/client.(*ProxyClient).ConnectToNode(0xc000712400, 0x5003a60, 0xc0003d4cc0, 0xc0002e6db0, 0x9, 0x4e3289b, 0x7, 0xc0002e6ce8, 0x8, 0x7ffeefbffb0c, ...)
	/private/tmp/teleport-20200507-15685-1a7mhlb/teleport-4.2.9/src/github.com/gravitational/teleport/lib/client/client.go:624 +0xe9e
github.com/gravitational/teleport/lib/client.(*TeleportClient).SSH(0xc00073e000, 0x5003a60, 0xc0003d4cc0, 0x0, 0x0, 0x0, 0xc00012e000, 0x0, 0x0)
	/private/tmp/teleport-20200507-15685-1a7mhlb/teleport-4.2.9/src/github.com/gravitational/teleport/lib/client/api.go:953 +0x23d
main.onSSH.func1(0x0, 0x0)
	/private/tmp/teleport-20200507-15685-1a7mhlb/teleport-4.2.9/src/github.com/gravitational/teleport/tool/tsh/tsh.go:787 +0x6c
github.com/gravitational/teleport/lib/client.RetryWithRelogin(0x5003a60, 0xc0003d4cc0, 0xc00073e000, 0xc0007879c8, 0x0, 0x1)
	/private/tmp/teleport-20200507-15685-1a7mhlb/teleport-4.2.9/src/github.com/gravitational/teleport/lib/client/api.go:314 +0x3c
main.onSSH(0xc000313440)
	/private/tmp/teleport-20200507-15685-1a7mhlb/teleport-4.2.9/src/github.com/gravitational/teleport/tool/tsh/tsh.go:786 +0xfa
main.Run(0xc000134010, 0x6, 0x6, 0x0)
	/private/tmp/teleport-20200507-15685-1a7mhlb/teleport-4.2.9/src/github.com/gravitational/teleport/tool/tsh/tsh.go:344 +0x4f29
main.main()
	/private/tmp/teleport-20200507-15685-1a7mhlb/teleport-4.2.9/src/github.com/gravitational/teleport/tool/tsh/tsh.go:185 +0xdd
```

### Debug

I already debugged into it. `proxy.teleportClient.localAgent` at the given line in stacktrace is nil. There are only two occasions when localAgent is set:
```go
	// sometimes we need to use external auth without using local auth
	// methods, e.g. in automation daemons
	if c.SkipLocalAuth {
		if len(c.AuthMethods) == 0 {
			return nil, trace.BadParameter("SkipLocalAuth is true but no AuthMethods provided")
		}
		// if the client was passed an agent in the configuration and skip local auth, use
		// the passed in agent.
		if c.Agent != nil {
			tc.localAgent = &LocalKeyAgent{Agent: c.Agent}
		}
	} else {
		// initialize the local agent (auth agent which uses local SSH keys signed by the CA):
		webProxyHost, _ := tc.WebProxyHostPort()
		tc.localAgent, err = NewLocalAgent(c.KeysDir, webProxyHost, c.Username, c.UseLocalSSHAgent)
		if err != nil {
			return nil, trace.Wrap(err)
		}
		if tc.HostKeyCallback == nil {
			tc.HostKeyCallback = tc.localAgent.CheckHostSignature
		}
	}
```
`c.SkipLocalAuth`is set to true as an identity file is given but `c.Agent` is nil so `tc.localAgent` is not set. I can't debug any further as I don't find the point where c.Agent should've been set in `tsh` the webclient seems to do it but not tsh.

--- a/lib/client/client.go
+++ b/lib/client/client.go
@@ -587,6 +587,9 @@ func (proxy *ProxyClient) ConnectToNode(ctx context.Context, nodeAddress NodeAdd
 	// creates a new context which holds the agent. if ForwardToAgent returns an error
 	// "already have handler for" we ignore it.
 	if recordingProxy {
+		if proxy.teleportClient.localAgent == nil {
+			return nil, trace.BadParameter("cluster is in proxy recording mode and requires agent forwarding for connections, but no agent was initialized")
+		}
 		err = agent.ForwardToAgent(proxy.Client, proxy.teleportClient.localAgent.Agent)
 		if err != nil && !strings.Contains(err.Error(), "agent: already have handler for") {
 			return nil, trace.Wrap(err)
@@ -672,6 +675,9 @@ func (proxy *ProxyClient) PortForwardToNode(ctx context.Context, nodeAddress Nod
 	// creates a new context which holds the agent. if ForwardToAgent returns an error
 	// "already have handler for" we ignore it.
 	if recordingProxy {
+		if proxy.teleportClient.localAgent == nil {
+			return nil, trace.BadParameter("cluster is in proxy recording mode and requires agent forwarding for connections, but no agent was initialized")
+		}
 		err = agent.ForwardToAgent(proxy.Client, proxy.teleportClient.localAgent.Agent)
 		if err != nil && !strings.Contains(err.Error(), "agent: already have handler for") {
 			return nil, trace.Wrap(err)

--- a/lib/client/interfaces.go
+++ b/lib/client/interfaces.go
@@ -139,7 +139,7 @@ func (k *Key) CertPrincipals() ([]string, error) {
 
 // AsAgentKeys converts client.Key struct to a []*agent.AddedKey. All elements
 // of the []*agent.AddedKey slice need to be loaded into the agent!
-func (k *Key) AsAgentKeys() ([]*agent.AddedKey, error) {
+func (k *Key) AsAgentKeys() ([]agent.AddedKey, error) {
 	// unmarshal certificate bytes into a ssh.PublicKey
 	publicKey, _, _, _, err := ssh.ParseAuthorizedKey(k.Cert)
 	if err != nil {
@@ -157,8 +157,8 @@ func (k *Key) AsAgentKeys() ([]*agent.AddedKey, error) {
 
 	// On Windows, return the certificate with the private key embedded.
 	if runtime.GOOS == teleport.WindowsOS {
-		return []*agent.AddedKey{
-			&agent.AddedKey{
+		return []agent.AddedKey{
+			{
 				PrivateKey:       privateKey,
 				Certificate:      publicKey.(*ssh.Certificate),
 				Comment:          comment,
@@ -180,15 +180,15 @@ func (k *Key) AsAgentKeys() ([]*agent.AddedKey, error) {
 	//
 	// For more details see the following: https://bugzilla.mindrot.org/show_bug.cgi?id=2550
 	// WARNING: callers expect the returned slice to be __exactly as it is__
-	return []*agent.AddedKey{
-		&agent.AddedKey{
+	return []agent.AddedKey{
+		{
 			PrivateKey:       privateKey,
 			Certificate:      publicKey.(*ssh.Certificate),
 			Comment:          comment,
 			LifetimeSecs:     0,
 			ConfirmBeforeUse: false,
 		},
-		&agent.AddedKey{
+		{
 			PrivateKey:       privateKey,
 			Certificate:      nil,
 			Comment:          comment,

--- a/lib/client/keyagent.go
+++ b/lib/client/keyagent.go
@@ -146,7 +146,7 @@ func (a *LocalKeyAgent) LoadKey(key Key) (*agent.AddedKey, error) {
 	// iterate over all teleport and system agent and load key
 	for _, agent := range agents {
 		for _, agentKey := range agentKeys {
-			err = agent.Add(*agentKey)
+			err = agent.Add(agentKey)
 			if err != nil {
 				a.log.Warnf("Unable to communicate with agent and add key: %v", err)
 			}
@@ -155,7 +155,7 @@ func (a *LocalKeyAgent) LoadKey(key Key) (*agent.AddedKey, error) {
 
 	// return the first key because it has the embedded private key in it.
 	// see docs for AsAgentKeys for more details.
-	return agentKeys[0], nil
+	return &agentKeys[0], nil
 }
 
 // UnloadKey will unload key for user from the teleport ssh agent as well as

--- a/tool/tsh/tsh.go
+++ b/tool/tsh/tsh.go
@@ -31,6 +31,7 @@ import (
 	"time"
 
 	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
 
 	"github.com/gravitational/teleport"
 	"github.com/gravitational/teleport/lib/asciitable"
@@ -1041,6 +1042,20 @@ func makeClient(cf *CLIConf, useProfileLogin bool) (*client.TeleportClient, erro
 			return nil, trace.Wrap(err)
 		}
 		c.AuthMethods = []ssh.AuthMethod{identityAuth}
+
+		// Also create an in-memory agent to hold the key. If cluster is in
+		// proxy recording mode, agent forwarding will be required for
+		// sessions.
+		c.Agent = agent.NewKeyring()
+		agentKeys, err := key.AsAgentKeys()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		for _, k := range agentKeys {
+			if err := c.Agent.Add(k); err != nil {
+				return nil, trace.Wrap(err)
+			}
+		}
 
 		if len(key.TLSCert) > 0 {
 			c.TLS, err = key.ClientTLSConfig()

--- a/tool/tsh/tsh_test.go
+++ b/tool/tsh/tsh_test.go
@@ -213,6 +213,12 @@ func (s *MainTestSuite) TestMakeClient(c *check.C) {
 	c.Assert(tc, check.NotNil)
 	c.Assert(tc.Config.WebProxyAddr, check.Equals, proxyWebAddr.String())
 	c.Assert(tc.Config.SSHProxyAddr, check.Equals, proxyPublicSSHAddr.String())
+	c.Assert(tc.LocalAgent().Agent, check.NotNil)
+	// Client should have an in-memory agent with keys loaded, in case agent
+	// forwarding is required for proxy recording mode.
+	agentKeys, err := tc.LocalAgent().Agent.List()
+	c.Assert(err, check.IsNil)
+	c.Assert(len(agentKeys), check.Not(check.Equals), 0)
 }
 
 func (s *MainTestSuite) TestIdentityRead(c *check.C) {

